core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/ldap/UniqueIDReadOnlyLDAPUserStoreManager.java

```diff
@@ -4246,4 +4250,157 @@ private List<String> getListOfGroupNamesWithPrimaryGroup(String searchBase, Stri
+
+        if (StringUtils.isBlank(configuredFormat)) {
+            return false;
+        }
+
+        // Create a Pattern object with the regex that matches a bracketed segment without nested brackets.
+        Pattern pattern = Pattern.compile("\\[[^\\[\\]]+\\]");
+        Matcher matcher = pattern.matcher(configuredFormat);
+
+        int count = 0;
+        // Iterate over all matches found by the matcher.
+        while (matcher.find()) {
+            count++;
+            // If more than one match is found, the configuration is invalid.
+            if (count > 1) {
+                return false;
+            }
+        }
+
+        // Valid if zero or one match was found.
+        return count <= 1;
+    }
+
+    /**
+     * Infers the LDAP group timestamp format from a sample LDAP entry.
+     * The patterns checked here correspond to formats defined in DEFAULT_LDAP_TIME_FORMATS_PATTERN
+     * ("[uuuuMMddHHmmss[,SSS][.SSS]X][uuuuMMddHHmmss[,SS][.SS]X][uuuuMMddHHmm[,S][.S]X]").
+     *
+     * @param answer                 A NamingEnumeration containing LDAP search results.
+     * @param timestampAttributeName The LDAP attribute name containing the timestamp.
+     * @return The inferred timestamp format string, or null if not detected.
+     * @throws NamingException If an error occurs while accessing LDAP attributes.
+     */
+    private String inferLDAPTimestampFormat(NamingEnumeration<SearchResult> answer, String timestampAttributeName)
+            throws NamingException {
+
+        if (answer != null && answer.hasMore()) {
+            SearchResult sr = answer.next();
+            Attribute attr = sr.getAttributes().get(timestampAttributeName);
+            if (attr != null) {
+                String sampleTimestamp = (String) attr.get();
+                // Case 1: 14 digits with no fractional seconds, e.g. 20250130083725Z or with offset ±HHMM
+                if (sampleTimestamp.matches("^\\d{14}([-+]\\d{4}|Z)$")) {
+                    return "uuuuMMddHHmmssX";
+                }
+                // Case 2: 14 digits with 3-digit fraction (comma or period), e.g. 20250130083725,801Z
+                // or 20250130083725.801Z
+                else if (sampleTimestamp.matches("^\\d{14}[,\\.]\\d{3}([-+]\\d{4}|Z)$")) {
+                    return sampleTimestamp.contains(",") ? "uuuuMMddHHmmss,SSSX" : "uuuuMMddHHmmss.SSSX";
+                }
+                // Case 3: 14 digits with 2-digit fraction, e.g. 20250130083725,80Z or 20250130083725.80Z
+                else if (sampleTimestamp.matches("^\\d{14}[,\\.]\\d{2}([-+]\\d{4}|Z)$")) {
+                    return sampleTimestamp.contains(",") ? "uuuuMMddHHmmss,SSX" : "uuuuMMddHHmmss.SSX";
+                }
+                // Case 4: 12 digits (up to minute) with 1-digit fraction, e.g. 202501300837,8Z or 202501300837.8Z
+                else if (sampleTimestamp.matches("^\\d{12}[,\\.]\\d{1}([-+]\\d{4}|Z)$")) {
```

Comment:
For these also shall we use precompile constants to improve the performance?

`private static final Pattern PATTERN_1 = Pattern.compile("^\\d{14}([-+]\\d{4}|Z)$"); `


```
if (PATTERN_1.matcher(sampleTimestamp).matches()) {
         return "uuuuMMddHHmmssX";
}
```

Reply:
Resolved via https://github.com/wso2/carbon-kernel/pull/4239/commits/d5d8bb16b2e30761acd99516f9460e75ab1ee9b4